### PR TITLE
Add some reduction forward AD formulas

### PR DESCRIPTION
**Pull request by soulitzer**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #69956
* #69884
* #69727
* #69740
* #69690
* __->__ #69661

Differential Revision: [D33020601](https://our.internmc.facebook.com/intern/diff/D33020601)

### CODE REVIEW

File: tools/autograd/derivatives.yaml [hunk 1]
```diff
@@ -1179,9 +1179,11 @@
 
 - name: prod(Tensor self, *, ScalarType? dtype=None) -> Tensor
   self: prod_backward(grad, self.to(grad.scalar_type()), result)
+  result: (prod_backward(ones_like(result), self_p.to(result.scalar_type()), result) * self_t.conj()).sum().conj()
```
Comment (albanD) [inline tools/autograd/derivatives.yaml:1182]: could be one one and expand it.

File: torch/csrc/autograd/FunctionsManual.cpp [hunk 2]
```diff
@@ -983,6 +985,29 @@ Tensor var_backward(Tensor grad, const Tensor& self, c10::optional<IntArrayRef>
...
+  }
+  return all_dims;
+}
+
+Tensor var_jvp(const Tensor& self_t, const Tensor& self_p, c10::optional<IntArrayRef> dim_opt,
+    c10::optional<int64_t> correction_opt, bool keepdim) {
+  auto correction = correction_opt.value_or(1);
+  if (self_p.dim() == 0 || !dim_opt.has_value()) {
+
+    return var_backward(self_t.conj(), self_p, correction).sum(all_dims_less_than(self_t.dim()), keepdim).conj();
```
Comment (albanD) [inline torch/csrc/autograd/FunctionsManual.cpp:1004]: maybe view_as(result) or get the result.sizes() as input to this function?

**Review by albanD — APPROVED** [review]
looks good if CI is happy

**Comment by soulitzer** [comment]
@soulitzer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33020601).

**Comment by soulitzer** [comment]
@soulitzer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33020601).

**Comment by soulitzer** [comment]
@soulitzer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33020601).

**Comment by soulitzer** [comment]
@soulitzer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33020601).

**Comment by soulitzer** [comment]
@soulitzer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33020601).

**Comment by soulitzer** [comment]
@soulitzer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33020601).

**Review by albanD — APPROVED** [review]

**Comment by soulitzer** [comment]
@soulitzer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33020601).

**Pull request closed without merging**